src/BCFKeyValueData.cc

```diff
@@ -216,6 +380,46 @@ Status BCFKeyValueData::dataset_bcf(const string& dataset,
     return Status::OK();
 }
 
+// Search all the buckets that may hold records within the query range. The tricky
+// corner case is the bucket immediately before the beginning of the query range.
+// It may hold records that start outside the range, but end inside it. We want
+// to return those as well.
+//
+// Return value: list of records that overlap with the query
+// range. This list may be empty, if no overlapping records were
+// found. If an invalid rid is requested, the return value will be an
+// empty list.
+Status BCFKeyValueData::dataset_bcf(const string& dataset,
+                                    const bcf_hdr_t* hdr,
+                                    const range& query,
+                                    vector<shared_ptr<bcf1_t> >& records) const {
+    Status s;
+    records.clear();
+
+    // basic sanity checks
+    if (query.rid < 0 || query.beg < 0 || query.end < 0)
+        return Status::NotFound();
```

Comment:
Let's make this `Status::Invalid("BCFKeyValueData::dataset_bcf: invalid query range", query.str())`. We should make error messages uniquely identifiable whenever possible since we won't have stack traces.  LGTM to squash and merge otherwise. Thanks!